# Requierements
```{r}
library(DESeq2)
library(tidyr)
library(dplyr)
library(edgeR)
library(shiny)
library(pasilla)
```

```{r "render README.Rmd"}


#  
#  data <- DT::datatable(iris)
#  data |>
#    DT::formatStyle("Species",
#                    backgroundColor = DT::styleEqual(unique(iris$Species),
#                                                     values = c("gray","yellow","#76D7C4")
#                                                     ))
#  set.seed(123)
#  colors <- c("#F08080","#E9967A","#DC143C",
#                   "#FFC0CB","#FF69B4","#DB7093",
#                   "#FFA07A","#FF6347","#FF8C00",
#                   "#FFFF00","#FFFACD","#FFDAB9",
#                   "#D8BFD8","#EE82EE","#BA55D3",
#                   "#ADFF2F","#98FB98","#3CB371",
#                   "#00FFFF","#E0FFFF","#7FFFD4",
#                   "#48D1CC","#5F9EA0","#87CEFA	")
#  sample(list_colors,(2+2),replace = TRUE,)
#  
#  rmarkdown::render("~/Documentos/R/my_pkgs/DEasy/README.Rmd",output_format = "html_document")
#  rmarkdown::render("~/Documentos/R/my_pkgs/DEasy/README.Rmd",output_format = "github_document")
```

```{r "import data example form package"}
## import data form package
pasCts <- system.file("extdata",
                      "pasilla_gene_counts.tsv",
                      package="pasilla", mustWork=TRUE)
pasAnno <- system.file("extdata",
                       "pasilla_sample_annotation.csv",
                       package="pasilla", mustWork=TRUE)
cts <- read.csv(pasCts,sep="\t",row.names="gene_id") |> as.matrix()
#cts <- cts[1:3000,] %>% as.matrix()
coldata <- read.csv(pasAnno, row.names=1)
coldata <- coldata[,c("condition","type")]
rownames(coldata) <- sub("fb", "", rownames(coldata)) ## coldata y count deben tner los minos nombre
all(rownames(coldata) %in% colnames(cts))
all(rownames(coldata) == colnames(cts))
cts <- cts[, rownames(coldata)]

```

# DESeq2

```{r "DESeq2 DE"}

validate_row_cols  <- function(df_s, df_r) {
  if (identical(row.names(df_s), colnames(df_r)) == FALSE) {
    stop("ERROR: Sample IDs on sample information rows and Raw counts columns must be the same \n see example data. ")
  }
}


DE_DESeq2_design  <- function(countData,
                              colData,
                              treatment,
                              alpha,
                              threshold,
                              interac = NULL) {
  countData <- as.matrix(countData)
  treatment =  treatment
  dds <- DESeq2::DESeqDataSetFromMatrix(countData = countData,
                                        colData = colData,
                                        design = treatment)
  
  keep <- rowSums(counts(dds)) >= 10
  dds <- dds[keep, ]
  dds <- DESeq(dds)
  res <- results(dds, alpha = alpha, lfcThreshold = 0)
  return(as.data.frame(res))
}

DE_DESeq2_main <- function(
                 countData,
                 colData,
                 treatment,
                 interac = NULL,
                 alpha,
                 threshold) {
  
  validate_row_cols(df_s =  colData,
                    df_r = countData)
  
  countData <- as.matrix(countData)
  
  if (is.null(interac)) {
    
    treatment =  as.formula(paste0("~ ", treatment))
    
    res <- DE_DESeq2_design(countData =  countData,
                    colData =   colData,
                    treatment =  treatment,
                    alpha = alpha,
                    threshold = threshold)
    
  } else if (interac == "None") {
    
    treatment =  as.formula(paste0("~ ", treatment))
    res <- DE_DESeq2_design(countData =  countData,
                            colData =   colData,
                            treatment =  treatment,
                            alpha = alpha,
                            threshold = threshold)
  } else {
    treatment =  as.formula(paste0("~ ", treatment, "+ ", interac))
    res <- DE_DESeq2_design(countData =  countData,
                            colData =   colData,
                            treatment =  treatment,
                            interac =  interac,
                            alpha = alpha,
                            threshold = threshold)
    
  }
  res <- as.data.frame(res)
  
  res <- res %>%
    mutate(., Difference = (
      case_when(
        .data$log2FoldChange >= threshold & .data$padj <= alpha ~ "UP",
        .data$log2FoldChange <= -threshold &
          .data$padj <= alpha ~ "DOWN",
        .data$log2FoldChange <= threshold |
          .data$padj > alpha ~ "Not significant",
        is.na(.data$padj) ~ "Not significant"
      )
    ))
  return(res)
}

```

```{r "DE with DESEQ2 main function examples"}

dds <- DESeqDataSetFromMatrix(countData = cts,
                              colData = coldata,
                              design = ~ condition)
dds
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
dds <- DESeq(dds)
res <- results(dds,alpha = 0.05,lfcThreshold = 0)
res


ix <- DE_DESeq2_main(
  countData = cts,
  colData = coldata,
  treatment = "condition",
  alpha = 0.05,
  threshold = 0,
  interac = "None"
)

sample_information <- read.csv("~/Documentos/R/my_pkgs/DEasy/tests/Samples_information.csv", row.names = 1)
raw_coutns <- read.csv("~/Documentos/R/my_pkgs/DEasy/tests/ara_counts.csv", row.names = 1)

iix <- DE_DESeq2_main(countData = raw_coutns,
                      colData = sample_information,
                      treatment = "SEX",
                      interac = "None",
                      alpha = 0.05,
                      threshold = 0)



```
# edgeR
```{r "edgeR DE"}


## edgeR DE_edgeR_main()

# load and filtering
dgList <- DGEList(counts = raw_coutns, genes = rownames(raw_coutns))
y <- dgList
design <- model.matrix(~ sample_information$SEX)
keep <- filterByExpr(y, design = design)
y <- y[keep,,keep.lib.sizes=FALSE]
y <- calcNormFactors(y) #
y <- estimateDisp(y, design, robust = TRUE)
fit <- glmQLFit(y, design)
qlf <- glmQLFTest(fit, coef = 2)
qlf <- p.adjust(qlf$table$FDR, method = "BH")


get_model <- function(df,var) {
  name <- quote(df) |> as.character()
  formula <- paste0("~", name,"$", as.character(var)) |> as.formula()
  model <- model.matrix(formula)
  return(model)
}

DE_edgeR_design <- function(countData,
                            colData,
                            treatment) {
  browser()
  dgList <- DGEList(counts = countData, genes = rownames(countData))
  y <- dgList
  design <- get_model(colData, treatment)
  keep <- filterByExpr(y, design = design)
  y <- y[keep,,keep.lib.sizes=FALSE]
  y <- calcNormFactors(y) #
  y <- estimateDisp(y, design, robust = TRUE)
  fit <- glmQLFit(y, design)
  qlf <- glmQLFTest(fit, coef = 2)
  qlf <- p.adjust(qlf$table$FDR, method = "BH")
  return(qlf)
}

DE_edgeR_main <- function(countData,
                          colData,
                          interac = NULL,
                          treatment) {
  browser()
  validate_row_cols(df_s =  colData,
                    df_r = countData)
  if (is.null(interac)) {
    
    treatment =  as.formula(paste0("~ ", treatment))
    qlf <- DE_edgeR_design(countData = countData,
                    colData = colData,
                    treatment = treatment)
    
  } else if (interac == "None") {
    treatment =  as.formula(paste0("~ ", treatment))
    qlf <- DE_edgeR_design(countData = countData,
                           colData = colData,
                           treatment = treatment)
    
  } else {
    treatment =  as.formula(paste0("~ ", treatment, "+ ", interac))
    qlf <- DE_edgeR_design(countData = countData,
                           colData = colData,
                           treatment = treatment)
    
    
  }
  return(qlf)
}
```

```{r "DE with edgeR main function examples"}
x <-  DE_edgeR_main(
  countData = raw_coutns,
  colData = sample_information,
  interac = "None",
  treatment = "SEX"
)
```

# Plots

## MDS
```{r "MDS"}
## MDS plot function
MDS_plot <- function(countData,
                     colData,
                     treatment,
                     interac = "None") {
  
  if (interac == "None") {
    design <- as.formula(paste0("~ ", treatment))
  } else if (is.null(interac)) {
    design <- as.formula(paste0("~ ", treatment))
  } else {
    design =  as.formula(paste0("~ ", treatment, "+ ", interac))
  }
  dds <- DESeqDataSetFromMatrix(countData = as.matrix(countData),
                                colData = colData,
                                design = design)
  vsd_0 <- vst(dds, blind = F) # calcualte dispersion trend
  sampleDists <- dist(t(assay(vsd_0))) #Calculate distance matrix
  sampleDistMatrix <- as.matrix( sampleDists ) # Create distance matrix
  mdsData <- data.frame(cmdscale(sampleDistMatrix)) #perform MDS
  mds <- cbind(mdsData, as.data.frame(colData(vsd_0)))
  
  F_vr_M_DESeq2_MDS <-  ggplot(mds, aes(X1,X2,color=SEX)) +
    geom_label_repel(aes(label = rownames(mds)), size = 3) +
    geom_point(size=3) +
    scale_color_manual(values =  c("#B22222","#8B008B"),
                       labels = c("Female", "Male"),
                       name = "Sex") +
    labs(title = "Females vr Males DESeq2",
         x = "Dim 1",
         y = "Dim 2") +
    theme_classic2()
  
  
  return(F_vr_M_DESeq2_MDS)
}


```
## Individual count plot funtion

```{r}

```

## heatmap plot function
```{r}

```

## MA plot 

```{r}

volcano_plot  <- function(DE_results, x, y, color) {
  plot <- ggplot(data = DE_results, aes(
    x = log2(baseMean),
    y = log2FoldChange,
    color = Difference
  )) + geom_point() +
        labs(color = "Differentially expressed",
             fill = "Number of transcripts")+ 
  xlab("Log2 Base-mean")+ ylab("log2 Fold-change")+
        theme_classic2() +
  # theme(axis.text.x = element_text(size = 18),
  #       axis.text.y = element_text(size = 18),
  #       axis.title.x = element_text(size = 20),
  #       axis.title.y = element_text(size = 20))+
  #       coord_fixed(ratio = 2/1) +
  #       scale_x_continuous(breaks = seq(from = 0, to = 20,2))+
  #       scale_y_continuous(breaks = seq(from = -10, to = 10,2)) +
        coord_flip()
  
  return(plot)
}


# volcanoplot_phoneutria
# ggsave(filename = "~/Documentos/R/R_Transcriptome/Transcriptome_Spider/Results_DE/Figure 3.pdf",device = "pdf")

volcano_plot(
  DE_results = as.data.frame(res),
  x = log2(.env$baseMean),
  y = .env$log2FoldChange,
  color = .env$Difference
)

```

## visual test

```{r}
dds <- DESeqDataSetFromMatrix(countData = cts,
                              colData = coldata,
                              design = ~ condition)
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
dds <- DESeq(dds)
res <- results(dds)
res
View(as.data.frame(head(res)))

res2 <- DE_DESeq2_main(countData = cts,
                       colData = coldata, 
                       treatment = "condition",interac = NULL,
                       alpha = 0.1,
                       threshold = 0)
View(head(res2))

MA_plot  <- function(df) {
  plot <- ggplot(data = df, 
                 aes(x = log2(baseMean),
                     y = log2FoldChange, 
                     color = Difference))  +
    geom_point() +
#        geom_hex(bins = 30) +
        labs(color = "Differentially expressed",
             fill = "Number of transcripts")+ 
  xlab("Log2 Base-mean")+ ylab("log2 Fold-change")+
        theme_classic2() +
        coord_flip()

  return(plot)
}

MA_plot(res2)
res2$Difference |> class()
```


